modules/renter/hostdb/negotiate.go

```diff
@@ -220,304 +119,207 @@ func (hu *hostUploader) negotiateContract(filesize uint64, duration types.BlockH
+		return hostContract{}, errors.New("couldn't read the contract signed by the host: " + err.Error())
 	}
 
 	// submit to blockchain
-	err = hu.hdb.tpool.AcceptTransactionSet(signedHostTxnSet)
+	err = tpool.AcceptTransactionSet(signedHostTxnSet)
 	if err == modules.ErrDuplicateTransactionSet {
 		// this can happen if the renter is uploading to itself
 		err = nil
 	}
 	if err != nil {
 		txnBuilder.Drop()
-		return err
-	}
-
-	hu.fcid = fcid
-	// create initial revision
-	hu.lastTxn.FileContractRevisions = []types.FileContractRevision{{
-		ParentID:              fcid,
-		UnlockConditions:      hu.unlockConditions,
-		NewRevisionNumber:     fc.RevisionNumber,
-		NewFileSize:           fc.FileSize,
-		NewFileMerkleRoot:     fc.FileMerkleRoot,
-		NewWindowStart:        fc.WindowStart,
-		NewWindowEnd:          fc.WindowEnd,
-		NewValidProofOutputs:  []types.SiacoinOutput{fc.ValidProofOutputs[0], fc.ValidProofOutputs[1]},
-		NewMissedProofOutputs: []types.SiacoinOutput{fc.MissedProofOutputs[0], fc.MissedProofOutputs[1]},
-		NewUnlockHash:         fc.UnlockHash,
-	}}
-
-	hu.hdb.mu.Lock()
-	hu.hdb.contracts[fcid] = hostContract{
+		return hostContract{}, err
+	}
+
+	// create host contract
+	hc := hostContract{
+		IP:           addr,
 		ID:           fcid,
 		FileContract: fc,
 		LastRevisionTxn: types.Transaction{
-			// first revision is empty
-			FileContractRevisions: []types.FileContractRevision{{}},
+			FileContractRevisions: []types.FileContractRevision{{
+				ParentID:              fcid,
+				UnlockConditions:      uc,
+				NewRevisionNumber:     fc.RevisionNumber,
+				NewFileSize:           fc.FileSize,
+				NewFileMerkleRoot:     fc.FileMerkleRoot,
+				NewWindowStart:        fc.WindowStart,
+				NewWindowEnd:          fc.WindowEnd,
+				NewValidProofOutputs:  []types.SiacoinOutput{fc.ValidProofOutputs[0], fc.ValidProofOutputs[1]},
+				NewMissedProofOutputs: []types.SiacoinOutput{fc.MissedProofOutputs[0], fc.MissedProofOutputs[1]},
+				NewUnlockHash:         fc.UnlockHash,
+			}},
 		},
 		SecretKey: ourSK,
 	}
-	hu.hdb.save()
-	hu.hdb.mu.Unlock()
 
-	return nil
+	return hc, nil
 }
 
-// Upload revises an existing file contract with a host, and then uploads a
-// piece to it.
-func (hu *hostUploader) Upload(data []byte) (uint64, error) {
-	// only one revision can happen at a time
-	hu.revisionLock.Lock()
-	defer hu.revisionLock.Unlock()
-
-	// get old file contract from renter
-	hu.hdb.mu.RLock()
-	fc, exists := hu.hdb.contracts[hu.fcid]
-	height := hu.hdb.blockHeight
-	hu.hdb.mu.RUnlock()
-	if !exists {
-		return 0, errors.New("no record of contract to revise")
+// newContract negotiates an initial file contract with the specified host
+// and returns a hostContract. The contract is also saved by the HostDB.
+func (hdb *HostDB) newContract(host modules.HostSettings) (hostContract, error) {
+	// reject hosts that are too expensive
+	if host.Price.Cmp(maxPrice) > 0 {
+		return hostContract{}, errTooExpensive
 	}
```

Comment:
by "previous function" do you mean `negotiateContract`? That function doesn't have access to the host's price (just the total contract payout), so it would have to be passed as an additional parameter.
Perhaps better would be for the hostdb to not return expensive hosts from `randomHosts`?